### DISCUSSION TITLE
czg-cli failed to load global configuration $HOME/.czrc or $HOME/.cz.json

### DISCUSSION THREAD

**Discussion by 13OnTheCode** [Zhengqbbb/cz-git#132] [Q&A] [answered]
```
System:
    OS: Windows 10 10.0.22621
    CPU: (8) x64 Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz
    Memory: 10.80 GB / 31.94 GB
Binaries:
    Node: 20.3.1 - ~\AppData\Local\pnpm\node.EXE
    npm: 9.6.7 - ~\AppData\Local\pnpm\npm.CMD
    pnpm: 8.6.5 - ~\AppData\Local\pnpm\pnpm.EXE
npmPackages:
    czg: ^1.6.1 => 1.6.1
```

可能与 #42 #125 有关

我注意到在Windows平台上无法成功读取全局配置

通过阅读源码，加载配置的核心代码应该是 `clLoader`, `czLoader`, `aiLoader` 这三个函数

但是似乎只有 `aiLoader` 通过 `const cwd = os.homedir()` 去传递 $HOME 路径

调试信息截图:

![df621bd4d0572b54ade6f7a6ca1e7d48.png](https://i2.mjj.rip/2023/07/02/df621bd4d0572b54ade6f7a6ca1e7d48.png)

由于对于源码没有过多研究，所以我疑惑的点是：

为什么只有 `aiLoader` 传递 `os.homedir()` ?  是不是 `clLoader` 以及 `czLoader` 有什么其他的处理导致不能直接传递参数？

如果我的疑惑的多余的，那么我认为只需要在顶层的 `loader` 函数去判断 `options.cwd` 是否为 `undefined`

如果是的话就赋值 `os.homedir()` 即可，我本地测试了似乎是没问题的，能够正常读取到全局配置文件

请问我有遗漏了什么地方吗？这么做是否可行？如果没问题的话我很乐意 Pull

It may be related to #42 #125

I noticed that the global configuration cannot be successfully read on the Windows platform.

By reading the source code, the core functions responsible for loading the configuration seem to be `clLoader`, `czLoader`, and `aiLoader`

However, it seems that only `aiLoader` passes the $HOME path through `const cwd = os.homedir()`

Screenshot of the debugging information:

...

Since I haven't studied the source code in depth, my points of confusion are:

Why is `os.homedir()` only passed in `aiLoader`? Are there any other processes in `clLoader` and `czLoader` that prevent direct parameter passing?

If my doubts are unnecessary, then I believe we only need to check `options.cwd` in the top-level `loader` function to see if it is `undefined`

If it is, we can assign `os.homedir()` to it. I tested it locally, and it seems to work fine. It can successfully read the global configuration file

Am I missing something? Is this approach feasible? If there are no issues, I would be happy to submit a Pull Request

**Comment by 13OnTheCode** [comment]
[packages/@cz-git/plugin-loader/src/index.ts](https://github.com/13OnTheCode/cz-git/commit/49f247def5a2cbc28d93dc999b7506c1e7f4bf21)

这是我调整后的代码，改动就一行代码：
`const result = await resultFn(searchPath) ?? await resultFn(os.homedir())`

我的理解是当 `await resultFn(searchPath)` 结果为 null 的话就意味着用户未指定配置文件或项目根目录下找不到配置文件

所以加了个条件继续判断 $HOME 目录内是否存在配置文件，如果存在则赋值。

请问是否可行呢？

**Comment by Zhengqbbb** [comment]
感谢你的调试过程，因为我目前没有 Windows 电脑所以没很好的处理这个事情
> 为什么只有 aiLoader 传递 os.homedir. 

因为只有他的配置是只存在 `HOME` 目录中的固定路径，不需要搜寻

> 如果我的疑惑的多余的，那么我认为只需要在顶层的 loader 函数去判断 options.cwd 是否为 undefined
> 如果是的话就赋值 os.homedir() 即可

这里 czLoader 与 clLoader 都是依靠了 [`cosmiconfig`](https://github.com/cosmiconfig/cosmiconfig) 这个库去做搜索，递归向 `HOME` 目录进行搜索配置，所以理应我不需要处理 `homedir` 的事项，但现在看起来在 Windows 环境中可能他存在问题

`const result = await resultFn(searchPath) ?? await resultFn(os.homedir())`

目前来看要启动两次 cosmiconfig 搜索配置。对于这个问题我持开放状态 PR Welcome。你可以多测试一下，目前我手里没有 Windows 的测试资源，不能很好的 DEBUG `cosmiconfig`～ 🤗

**Comment by 13OnTheCode** [comment, marked as answer]
是的, 我在本地调试了 `cosmiconfig`, 确实是 `cosmiconfig` 的问题：

[https://github.com/cosmiconfig/cosmiconfig/issues/278](https://github.com/cosmiconfig/cosmiconfig/issues/278)
[https://github.com/Exlint/cz-vinyl/issues/71](https://github.com/Exlint/cz-vinyl/issues/71)

但是目前我还没有深入阅读源码寻找问题所在，初步测试发现应该是因为 MacOS 是统一的目录结构, 有一个统一的主目录（Home Directory）作为用户的个人文件空间，而在 Windows 上却是使用了分区的概念, 属于分层的目录结构, 通过逻辑磁盘（如C、D、E等）来组织文件和目录。每个逻辑磁盘都有自己的根目录。稍微看了一下源码 `cosmiconfig` 似乎没有对 Windows 系统做额外的处理，不清楚是不是开发人员有意为之还是疏忽了。

测试过程：

我在 `D:\Developer\13OnTheCode\test\index.js` 内执行了 `cosmiconfig(...).search(...)`
那么 `cosmiconfig` 最后递归到的 "主目录" 是 D 盘，而不是 C 盘
所以当 .czrc 配置文件创建在 D 盘的根目录下是可以找到的

-----

最后我提交了一个 PR #133 ，最终改动是 `const result = await resultFn(searchPath) ?? await cosmiconfigFn.search(os.homedir())` 

`cosmiconfigFn.search(os.homedir())` 因为指定了 C 盘主目录路径，所以我认为不会有性能上的问题，在 `cosmiconfig `修复这个问题之前，我认为这是目前最优解。稍后我会去阅读 `cosmiconfig` 源码，看看能不能直接从源头修复这个问题。

**Reply by Zhengqbbb** [reply]
噢～  你的项目时放在 D盘所以 `cosmiconfig` 才会递归到 D 盘根目录结束这个出发点是可以的，因为他不能跨盘递归
感谢你的 Debug 过程。我已经发布了 next 版本你可以进行测试 `1.6.2-beta.6`

```sh
npm install -g cz-git@next
```

正式版本还需要一段时间预计这两周，我需要进行一些调整后才会发布

**Reply by 13OnTheCode** [reply]
是的，我相信其他遇到无法读取全局配置的用户都是因为问题导致。
我正在咨询 cosmiconfig 开发团队人员是否愿意在 Windows 系统上做一个额外的处理来解决项目不在 C 盘的问题。
如果官方采纳我会回来更正本次修改。
感谢作者！

**Reply by Zhengqbbb** [reply]
czg 🌟(v1.7.0)🌟 版本已经发布，可以进行更新并使用，如果有其他问题可以及时和我沟通 🤗

**Comment by WakerCN** [comment]
# 环境
* czg 1.8
* 系统 windows
* 使用了nvm，配置了node全局包安装在d盘 `D:\develop\nodejs`

# 操作步骤
* npm install -g czg
* 配置~/.czrc文件,进行了自定义配置

使用czg命令时仍然无法读取到全局.czrc中的配置，并且尝试了在使用命令的同一个磁盘的根路径下放一份`.czrc`也没有生效

看到上面提的pr是为了修复windows系统下无法读取全局配置，下载目前最新版的czg好像没有生效

下面贴上我的`~/.czrc`文件
```json
{
  "path": "cz-git",
  "$schema": "https://cdn.jsdelivr.net/gh/Zhengqbbb/cz-git@1.8.0/docs/public/schema/cz-git.json",
  "rules": {},
  "prompt": {
    "alias": {
      "fd": "docs: fix typos"
    },
    "messages": {
      "type": "Select the type of change that you're committing:",
      "scope": "Denote the SCOPE of this change (optional):",
      "customScope": "Denote the SCOPE of this change:",
      "subject": "Write a SHORT, IMPERATIVE tense description of the change:\n",
      "body": "Provide a LONGER description of the change (optional). Use \"|\" to break new line:\n",
      "breaking": "List any BREAKING CHANGES (optional). Use \"|\" to break new line:\n",
      "footerPrefixesSelect": "Select the ISSUES type of changeList by this change (optional):",
      "customFooterPrefix": "Input ISSUES prefix:",
      "footer": "List any ISSUES by this change. E.g.: #31, #34:\n",
      "generatingByAI": "Generating your AI commit subject...",
      "generatedSelectByAI": "Select suitable subject by AI generated:",
      "confirmCommit": "Are you sure you want to proceed with the commit above?"
    },
    "types": [
      {
        "value": "feat",
        "name": "feat:     ✨  A new feature",
        "emoji": ":sparkles:"
      },
      {
        "value": "fix",
        "name": "fix:      🐛  A bug fix",
        "emoji": ":bug:"
      },
      {
        "value": "docs",
        "name": "docs:     📝  Documentation only changes",
        "emoji": ":memo:"
      },
      {
        "value": "style",
        "name": "style:    💄  Changes that do not affect the meaning of the code",
        "emoji": ":lipstick:"
      },
      {
        "value": "refactor",
        "name": "refactor: ♻️   A code change that neither fixes a bug nor adds a feature",
        "emoji": ":recycle:"
      },
      {
        "value": "perf",
        "name": "perf:     ⚡️  A code change that improves performance",
        "emoji": ":zap:"
      },
      {
        "value": "test",
        "name": "test:     ✅  Adding missing tests or correcting existing tests",
        "emoji": ":white_check_mark:"
      },
      {
        "value": "build",
        "name": "build:    📦️   Changes that affect the build system or external dependencies",
        "emoji": ":package:"
      },
      {
        "value": "ci",
        "name": "ci:       🎡  Changes to our CI configuration files and scripts",
        "emoji": ":ferris_wheel:"
      },
      {
        "value": "chore",
        "name": "chore:    🔨  Other changes that don't modify src or test files",
        "emoji": ":hammer:"
      },
      {
        "value": "revert",
        "name": "revert:   ⏪️  Reverts a previous commit",
        "emoji": ":rewind:"
      }
    ],
    "useEmoji": true,
    "emojiAlign": "center",
    "useAI": false,
    "aiNumber": 1,
    "themeColorCode": "",
    "scopes": [],
    "allowCustomScopes": true,
    "allowEmptyScopes": true,
    "customScopesAlign": "bottom",
    "customScopesAlias": "custom",
    "emptyScopesAlias": "empty",
    "upperCaseSubject": false,
    "markBreakingChangeMode": false,
    "allowBreakingChanges": [
      "feat",
      "fix"
    ],
    "breaklineNumber": 100,
    "breaklineChar": "|",
    "skipQuestions": [],
    "issuePrefixes": [
      {
        "value": "closed",
        "name": "closed:   ISSUES has been processed"
      }
    ],
    "customIssuePrefixAlign": "top",
    "emptyIssuePrefixAlias": "skip",
    "customIssuePrefixAlias": "custom",
    "allowCustomIssuePrefix": true,
    "allowEmptyIssuePrefix": true,
    "confirmColorize": true,
    "maxHeaderLength": 100,
    "maxSubjectLength": 20,
    "minSubjectLength": 0,
    "defaultBody": "",
    "defaultIssues": "",
    "defaultScope": "",
    "defaultSubject": ""
  }
}
```
请问一下是我的操作步骤有问题嘛 @13OnTheCode

**Reply by Zhengqbbb** [reply]
如果在 D 盘根目录创建 `.czrc` 是否可以呢

**Reply by WakerCN** [reply]
好像不行耶，我尝试放在 `D:\.czrc` 或者 `D:\develop\nodejs\.czrc` 都读取不到全局配置文件

不知道有没有其他windows用户和我遇到一样的问题

**Reply by Zhengqbbb** [reply]
运行一下，下面的命令，把结果输出出来看看

```bash
$env:CZ_DEBUG = 1
czg
```

```bash
npx -y  envinfo --system --binaries --npmPackages cz-git --npmPackages commitizen --npmPackages czg --npmGlobalPackages cz-git --npmGlobalPackages commitizen --npmGlobalPackages czg
```

![CleanShot 2024-02-01 at 11 21 06@2x](https://github.com/Zhengqbbb/cz-git/assets/40693636/c04b281a-3e11-43e8-a611-4a6523410131)


---

我需要看看 `cz config-file` 读到哪里去了，最怕就是在项目根目录的 package.json 就停了，因为 `package.json` 当中的 `config.commitizen` 也算是 `czrc`

**Reply by WakerCN** [reply]
当前项目目录为 `F:\imooc-study\imooc-vue-react-study`
* 第一步：在此目录下打开终端输入
```bash
$env:CZ_DEBUG = 1
czg
```

* 第二步：在此目录下执行命令
```bash
npx -y  envinfo --system --binaries --npmPackages cz-git --npmPackages commitizen --npmPackages czg --npmGlobalPackages cz-git --npmGlobalPackages commitizen --npmGlobalPackages czg
```
输出结果：
![image](https://github.com/Zhengqbbb/cz-git/assets/50485199/20dc9db4-5d68-481b-9dc8-c87f8cab3541)

**Reply by Zhengqbbb** [reply]
不会是在 F 盘的根目录吧 🙃

**Reply by Zhengqbbb** [reply]
看样子应该是 `cosmiconfig` 这个配置加载库的问题，虽然在最新版有所更新修复。
但是在更新版本后 czg 的打包体积从 1.7 MB 到了 5 MB。我接受不了
